### PR TITLE
p2p: de-dupe and shuffle public rooms; implement keydb

### PR DESCRIPTION
**Pull request by Kegsay**
- We don't want a serverKeyAPI as fetching keys doesn't need a DB.
- De-dupe rooms so we don't see them multiple times, but shuffle the
  alias we join via so we don't all flood a single server.

### CODE REVIEW

File: cmd/dendritejs/main.go [hunk 1]
```diff
@@ -197,14 +196,12 @@ func main() {
...
 	federation := createFederationClient(cfg, node)
 
-	serverKeyAPI := serverkeyapi.SetupServerKeyAPIComponent(
-		base, federation,
-	)
+	fetcher := &libp2pKeyFetcher{}
 	keyRing := gomatrixserverlib.KeyRing{
 		KeyFetchers: []gomatrixserverlib.KeyFetcher{
-			&libp2pKeyFetcher{},
+			fetcher,
```
Comment (neilalexander) [inline cmd/dendritejs/main.go:202]: I think if you already have `fetcher` satisfying `KeyDatabase` then you probably don't need to specify any `KeyFetchers` at all.

Comment (Kegsay) [reply cmd/dendritejs/main.go:202]: Can't hurt to have it?

Comment (neilalexander) [reply cmd/dendritejs/main.go:202]: Won't do any harm, nope. It'll probably never get hit.

Comment (Kegsay) [reply cmd/dendritejs/main.go:202]: That's what she said

**Pull request merged**